sw/host/opentitanlib/src/spiflash/sfdp.rs

```diff
@@ -0,0 +1,503 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+use byteorder::{LittleEndian, ReadBytesExt};
+use num_enum::FromPrimitive;
+use serde::Serialize;
+use std::convert::TryFrom;
+use thiserror::Error;
+
+use crate::util::bitfield::BitField;
+
+#[derive(Debug, Error)]
+pub enum Error {
+    #[error(transparent)]
+    Io(#[from] std::io::Error),
+    #[error(transparent)]
+    TryFromIntError(#[from] std::num::TryFromIntError),
+    #[error("the range {0}..{1} is out of bounds")]
+    SliceRange(usize, usize),
+
+    // This is only needed to meet the error conversion requirements for the most
+    // general case in the field! macro below.
+    #[error(transparent)]
+    Infallible(#[from] std::convert::Infallible),
+}
+
+#[derive(Debug, Serialize)]
+/// The SFDP header identifies a valid SFDP, its version and the number of parameter headers.
+pub struct SfdpHeader {
+    pub signature: u32,
+    pub minor: u8,
+    pub major: u8,
+    pub nph: u8,
+    pub reserved: u8,
+}
+
+impl TryFrom<&[u8]> for SfdpHeader {
+    type Error = Error;
+    fn try_from(buf: &[u8]) -> Result<Self, Self::Error> {
+        let mut reader = std::io::Cursor::new(buf);
+        Ok(SfdpHeader {
+            signature: reader.read_u32::<LittleEndian>()?,
+            minor: reader.read_u8()?,
+            major: reader.read_u8()?,
+            nph: reader.read_u8()?,
+            reserved: reader.read_u8()?,
+        })
+    }
+}
+
+#[derive(Debug, Serialize)]
```

Comment:
Nit: be consistent: `#[attr]`s before or after doc comments?

Reply:
heh.  Again, something a formatting tool should do for me.